internal/common/spiretrustsource.go

```diff
@@ -0,0 +1,192 @@
+		}
+
+		source.spireEndpoints[trustDomain] = &SpireEndpoint{
+			domain:    trustDomain,
+			loadState: Pending,
+			spireURL:  spireURL,
+			client:    client,
+		}
+	}
+
+	for _, spireEndpoint := range source.spireEndpoints {
+		if err := spireEndpoint.client.Start(); err != nil {
+			return nil, fmt.Errorf("failed to start NewX509SVIDClient for domain %s: %v", spireEndpoint.domain, err)
+		}
+	}
+
+	return source, nil
+}
+
+// Stop stops all spire clients
+func (s *SpireTrustSource) Stop() error {
+	errs := make([]string, 0)
+	for _, spireEndpoint := range s.spireEndpoints {
+		if err := spireEndpoint.client.Stop(); err != nil {
+			errs = append(errs, fmt.Sprintf("domain %s: %v", spireEndpoint.domain, err))
+		}
+	}
+
+	if len(errs) != 0 {
+		return fmt.Errorf("failed to stop NewX509SVIDClients: (%s)", strings.Join(errs, "),("))
+	}
+
+	return nil
+}
+
+func (w *workloadWatcher) UpdateX509SVIDs(svids *workload.X509SVIDs) {
+	certs := svids.Default().TrustBundle
+	w.source.domainCertificates[w.domain] = certs
+	w.source.spireEndpoints[w.domain].loadState = Loaded
+
+	if w.localStoragePath != "" {
+		builder := strings.Builder{}
+		for _, cert := range certs {
+			block := &pem.Block{
+				Type:  "CERTIFICATE",
+				Bytes: cert.Raw,
+			}
+			builder.Write(pem.EncodeToMemory(block))
+		}
+		file, err := appFS.OpenFile(w.localStoragePath, os.O_WRONLY|os.O_CREATE, 0600)
+		if err != nil {
+			logrus.Warnf("could not open backup file for trust domain %s at %s: %v", w.domain, w.localStoragePath, err)
+		} else {
+			defer file.Close()
+			_, err = file.WriteString(builder.String())
+			if err != nil {
+				logrus.Warnf("could not write to backup file for trust domain %s at %s: %v", w.domain, w.localStoragePath, err)
+			}
+		}
+	}
+
+	select {
+	case w.source.updateChan <- struct{}{}:
+	case <-time.After(w.source.updateTimeout):
+	}
+}
+
+func (w *workloadWatcher) OnError(err error) {
+	if w.source.spireEndpoints[w.domain].loadState == Pending {
+		if w.localStoragePath != "" {
+			domainPaths := map[string][]string{
+				w.domain: []string{w.localStoragePath},
+			}
+
+			if fileTrustSource, err := NewFileTrustSource(domainPaths); err != nil {
+				logrus.Warnf("could not load certs for domain %s from disk: %v", w.domain, err)
+				w.source.spireEndpoints[w.domain].loadState = Failed
+			} else {
+				w.source.domainCertificates[w.domain] = fileTrustSource.TrustedCertificates()[w.domain]
+				w.source.spireEndpoints[w.domain].loadState = LoadedFromBackup
+				logrus.Infof("loaded certs for domain %s from disk", w.domain)
+			}
+		} else {
+			w.source.spireEndpoints[w.domain].loadState = Failed
+			logrus.Warnf("could not connect to spire server for domain %s and local storage disabled", w.domain)
+		}
+	} else {
+		// if the state was already Loaded, LoadedFromBackup, or Failed then don't do anything
+		w.source.spireEndpoints[w.domain].loadState = Failed
+	}
```

Comment:
If the state was `Loaded` or `LoadedFromBackup` then we're setting the state to `Failed`. This isn't in alignment with the comment, and I think is wrong. If we're already `Loaded` or `LoadedFromBackup` then a failure could (?should?) be logged, but the already loaded certs are still usable and thus the state should stay as it was.

Reply:
Yeah you're right on this.